src/backend/utils/resgroup/resgroup.c

```diff
@@ -4766,6 +4796,78 @@ moveQueryCheck(int sessionId, Oid groupId)
 	cdbdisp_clearCdbPgResults(&cdb_pgresults);
 }
 
+/*
+ * Try to give away all slot control to target process.
+ */
+static bool
+resGroupGiveSlotAway(int sessionId, void *slot, Oid groupId)
+{
+	long commandId;
+	bool moved = false;
+	
+	/* 
+	 * Be sure handler will set latch for the right command.
+	 * It's important to generate new command ID every time. If we'll get
+	 * an error somewhere after sending signal, e.g. on WaitLatch, the
+	 * next call to ResGroupMoveQuery will not geenerate new ID, so old
+	 * handler can set our latch.
+	 */
+	commandId = random() % MAX_COMMAND_ID;
+
+	SpinLockAcquire(&MyProc->movetoMutex);
+	MyProc->movetoCommandId = commandId;
+	SpinLockRelease(&MyProc->movetoMutex);
+	ResetLatch(&MyProc->procLatch);
+
+	SIMPLE_FAULT_INJECTOR("resource_group_move_signal1");
+
+	if (!ResGroupSignalMoveQuery(sessionId, slot, InvalidOid, commandId))
+		elog(ERROR, "can't send signal; is target process alive?");
+	
+	PG_TRY();
+	{
+		SIMPLE_FAULT_INJECTOR("resource_group_move_signal2");
+		/*
+		 * We don't analyze returned WaitLatch state, because we will get the
+		 * actual result by changed (or not) movetoCommandId value. There is
+		 * no much difference of why process wasn't moved (by timeout or by
+		 * death reasons), but the logic around became more complex.
+		 */
+		WaitLatch(&MyProc->procLatch,
```

Comment:
Curious to know do we have some pitfalls not analyzing `WL_POSTMASTER_DEATH` return status like it's done [here](https://github.com/arenadata/gpdb/blob/c1a3ea3c2af5d329fdb012abc18c8ca42f5612b7/src/backend/utils/resgroup/resgroup.c#L2940).